espnet2/uasr/espnet_model.py

```diff
@@ -0,0 +1,476 @@
+import logging
+import argparse
+
+from contextlib import contextmanager
+from typing import Dict, List, Optional, Tuple, Union
+
+import torch
+import kenlm
+import editdistance
+
+import torch.nn.functional as F
+from packaging.version import parse as V
+from typeguard import check_argument_types
+
+from espnet.nets.pytorch_backend.nets_utils import make_pad_mask
+from espnet2.text.token_id_converter import TokenIDConverter
+
+from espnet2.asr.frontend.abs_frontend import AbsFrontend
+from espnet2.uasr.discriminator.abs_discriminator import AbsDiscriminator
+from espnet2.uasr.generator.abs_generator import AbsGenerator
+from espnet2.uasr.segmenter.abs_segmenter import AbsSegmenter
+
+from espnet2.uasr.loss.discriminator_loss import UASRDiscriminatorLoss
+from espnet2.uasr.loss.gradient_penalty import UASRGradientPenalty
+from espnet2.uasr.loss.phoneme_diversity_loss import UASRPhonemeDiversityLoss
+from espnet2.uasr.loss.pseudo_label_loss import UASRPseudoLabelLoss
+from espnet2.uasr.loss.smoothness_penalty import UASRSmoothnessPenalty
+
+from espnet2.torch_utils.device_funcs import force_gatherable
+from espnet2.train.abs_espnet_model import AbsESPnetModel
+from espnet2.utils.types import str2bool
+
+if V(torch.__version__) >= V("1.6.0"):
+    from torch.cuda.amp import autocast
+else:
+    # Nothing to do if torch<1.6.0
+    @contextmanager
+    def autocast(enabled=True):
+        yield
+
+
+class ESPnetUASRModel(AbsESPnetModel):
+    """Unsupervised ASR model"""
```

Comment:
We should somehow acknowledge the original code source from fairseq

Reply:
done